### PR TITLE
Donuts will now heal security and ERT personnel

### LINKED ISSUE #5427: Donuts should heal security personnel
## Description

This isn't a goof. In /tg/, donuts heal a small amount of brute/burn damage when eaten by security personnel. I believe it's 5 damage healed for most donuts, 10 for frosted and more for a couple rarer flavors.

### PR DESCRIPTION
**Pull request by MaxIsJoe**
### Purpose
_Describe the problem the PR fixes or the feature it introduces_<br>
As the title suggests, donuts will now heal brute damage for players who have a security or ERT role. This was suggested in #5427 .
closes #5427 

### Notes:

- I couldn't find the exact numbers or behavior for how this exactly works in the /tg/ base so I assumed from what #5427 said that donuts heal all body parts that have brute damage on them.
- This does not apply only to specific jobs since you can add job from the inspector to have this special interaction.

### Changelog:

- Donuts will heal players with specified jobs from a list.
- - A donut script has been created that inherits from edible to create this interaction.
- - The majority of donuts heal security for 5 brute damage, Jelly heals 10, slime jelly heals 12.

### CODE REVIEW

**Comment by corp-0** [comment]
fixes #5427 <- can you add that to your description?

**Comment by ThatDan123** [comment]
you can edit as well Gilles

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections;
+using UnityEngine;
+
+
+public class Donut : Edible
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]: You could simplify this component into a Heal job edible, so in the editor you can specify what jobs it heals for and how much

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+{
+    [Tooltip("The amount of brute damage healed for a player who has a secuirty role.")]
+    public int SecurityHealBruteDamage = 5;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]:
```suggestion
    [SerialiezedField]
    private int securityHealBruteDamage = 5;
```

We don't really need to know this number anywhere else. Make it private and add the ``SerializedField`` attribute.

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    /// <summary>
+	/// Checks if the player eating the donut has a secuirty job and heals them if they are part of security.
+    /// If the player is not part of secuirty then nothing special will be triggered.
+	/// </summary>
+	public override void Eat(PlayerScript eater, PlayerScript feeder)
+	{
+		if (CheckForJob(eater))
+		{
+            Heal(eater);
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]: seems like an extra level of indentation here

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+    /// <summary>
+	/// Heals all body parts that have brute damage.
+	/// </summary>
+    private void Heal(PlayerScript player)
+    {
+        var livingHealth = player.GetComponent<LivingHealthBehaviour>();
+        foreach (BodyPartBehaviour BodyPart in livingHealth.BodyParts)
+		{
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]: here indentation looks weird. Are you using spaces instead of tabs?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: In vscode it looks normal to me, and I use tabs yeah.
Working on fixing a lot of issues here now.

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                return true;
+            case JobType.DETECTIVE:
+                return true;
+            case JobType.WARDEN:
+                return true;
+            default:
+                return false;
+        }
+    }
+	private void Eat(PlayerScript eater, PlayerScript feeder, int nutrition)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]: Looks like this part wasn't altered at all. You could instead add your job check and heal in the public method and then call ``Base.Eat()``

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: So do I delete the override function and move everything in the public Eat()?

Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: let me inspect the base class.

Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: Ok I see what you did. Delete the private and call ``Base.Eat()`` in your overridden public **after** doing your heal check.

**Comment by PerfectTangent** [comment]
I suppose it should also effect ERT Security Officers. 

Also, in the future, it might be best to  make this system a quirk that can be attached to the occupation SOs so it isn't hardcoded, much like the mime's spells. It will allow this donut healing quirk to work cohesively with the other occupation quirks that we will likely add at some point. We want to move away from using the hardcoded JobTypes in the future anyways, don't we?

**Comment by corp-0** [comment]
Yes, I'd advocate to do what Bod suggested and make it a generic thing that can be set in inspector. Still my other suggestions are relevant.

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]:
```
public override void Eat(PlayerScript eater, PlayerScript feeder)
	{
		if (CheckForJob(eater)) 
		{
			Heal(eater);
		}
		base.Eat(eater, feeder);
	}
```
like this?

Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: yep, that should work.

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs (hunk 1 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: only thing missing is changing the name of the component so it is relevant for any other edible that heals a job.

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+
+public class Donut : Edible
+{
+	[SerializeField, Tooltip("The amount of brute damage healed for a player who has a specfied role.")]
+	private int healBruteDamage = 5;
+	public List<JobType> healableJobs;
+
+	/// <summary>
+	/// Checks if the player eating the donut has a secuirty job and heals them if they are part of security.
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/Donut.cs]: update the docstring

File: UnityProject/Assets/Scripts/Items/Food/Donut.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: what and where do I name it to again?

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: Rename the Class to something like HealJobEdible, and then you'll need to rename the file in unity to that as well (might need to replace the component on the donuts)

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Food/Donut.cs]: ooohhh, gotcha. Will do that once Unity stops crashing.

File: UnityProject/Assets/Scripts/Items/Food/HealableJobEdible.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
+public class HealableJobEdible : Edible
+{
+	[SerializeField, Tooltip("The amount of brute damage healed for a player who has a specfied role.")]
+	private int healBruteDamage = 5;
+	public List<JobType> healableJobs;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Food/HealableJobEdible.cs]: this public isn't used anywhere but in this class. Make it private with serialized field attribute just like the int for healing amount.

**Pull request merged**